src/nvim/event/socket.c

```diff
@@ -17,60 +17,46 @@
 #include "nvim/path.h"
 #include "nvim/memory.h"
 #include "nvim/macros.h"
+#include "nvim/charset.h"
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
 # include "event/socket.c.generated.h"
 #endif
 
-#define NVIM_DEFAULT_TCP_PORT 7450
-
-void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
-    const char *endpoint, void *data)
-  FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_NONNULL_ARG(2) FUNC_ATTR_NONNULL_ARG(3)
+int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+                        const char *endpoint)
+  FUNC_ATTR_NONNULL_ALL
 {
-  // Trim to `ADDRESS_MAX_SIZE`
-  if (xstrlcpy(watcher->addr, endpoint, sizeof(watcher->addr))
-      >= sizeof(watcher->addr)) {
-    // TODO(aktau): since this is not what the user wanted, perhaps we
-    // should return an error here
-    WLOG("Address was too long, truncated to %s", watcher->addr);
-  }
-
-  bool tcp = true;
-  char ip[16], *ip_end = xstrchrnul(watcher->addr, ':');
-
-  // (ip_end - addr) is always > 0, so convert to size_t
-  size_t addr_len = (size_t)(ip_end - watcher->addr);
-
-  if (addr_len > sizeof(ip) - 1) {
-    // Maximum length of an IPv4 address buffer is 15 (eg: 255.255.255.255)
-    addr_len = sizeof(ip) - 1;
-  }
-
-  // Extract the address part
-  xstrlcpy(ip, watcher->addr, addr_len + 1);
-  int port = NVIM_DEFAULT_TCP_PORT;
-
-  if (*ip_end == ':') {
-    // Extract the port
-    long lport = strtol(ip_end + 1, NULL, 10); // NOLINT
-    if (lport <= 0 || lport > 0xffff) {
-      // Invalid port, treat as named pipe or unix socket
-      tcp = false;
-    } else {
-      port = (int) lport;
+  xstrlcpy(watcher->addr, endpoint, sizeof(watcher->addr));
+  char *addr = watcher->addr;
+  char *host_end = strrchr(addr, ':');
+
+  if (host_end && addr != host_end) {
+    char *port = host_end + 1;
+    intmax_t iport;
+
+    int ret = getdigits_safe(&(char_u *){ (char_u *)port }, &iport);
+    if (ret == FAIL || iport < 0 || iport > UINT16_MAX) {
+      ELOG("Invalid port: %s", port);
+      return UV_EINVAL;
     }
-  }
 
-  if (tcp) {
-    // Try to parse ip address
-    if (uv_ip4_addr(ip, port, &watcher->uv.tcp.addr)) {
-      // Invalid address, treat as named pipe or unix socket
-      tcp = false;
+    *host_end = '\0';
+    uv_getaddrinfo_t request;
+
+    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
+                                &(struct addrinfo){
+                                  .ai_family = AF_UNSPEC,
+                                  .ai_socktype = SOCK_STREAM,
+                                });
+    *host_end = ':';
```

Comment:
How about we add a comment and remove `*host_end=':';` and `*port='\0';`? This code is not needed anymore, because we don't continue in the "pipe" part.
```diff
diff --git a/src/nvim/event/socket.c b/src/nvim/event/socket.c
index 8509bb1ea..3a47f3489 100644
--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -33,6 +33,9 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
   char *host_end = strrchr(addr, ':');
 
   if (host_end && addr != host_end) {
+    // Split user specified address into two strings, addr(hostname) and port.
+    // The port part in watcher->addr will be updated later.
+    *host_end = '\0';
     char *port = host_end + 1;
     intmax_t iport;
 
@@ -42,7 +45,6 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
-    *host_end = '\0';
     uv_getaddrinfo_t request;
 
     int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
@@ -50,12 +52,10 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
-    *host_end = ':';
     if (retval != 0) {
       ELOG("Host lookup failed: %s", addr);
       return retval;
     }
-    *port = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
 
     uv_tcp_init(&loop->uv, &watcher->uv.tcp.handle);
@@ -102,7 +102,7 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
           : ((struct sockaddr_in6 *)&sas)->sin6_port;
         // v:servername uses the string from watcher->addr
         size_t len = strlen(watcher->addr);
-        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, "%" PRIu16,
+        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, ":%" PRIu16,
                  ntohs(port));
         break;
       }

```